### PR TITLE
internal: Prepare parser interface for editions

### PR DESCRIPTION
**Pull request by Urhengulas**

### CODE REVIEW

**Comment by Veykril** [comment]
The one in `project-model` exists for rust-project.json deserialization, `base-db` is the representation for crates and the `parser` one is separate because it does not depend on `base-db` (and shouldn't). On another note, the `base-db` one we might want to move to the `span` crate, I think that makes more sense there. The parser might depend on the span crate in the future in which case we can collapse them, but right now that makes little sense

**Comment by Veykril** [comment]
Rebased and deduplicated the enum

**Comment by Veykril** [comment]
Thanks!
@bors r+

File: crates/hir-ty/src/method_resolution.rs [hunk 1]
```diff
@@ -1157,7 +1157,7 @@ fn iterate_trait_method_candidates(
         {
             // FIXME: this should really be using the edition of the method name's span, in case it
             // comes from a macro
-            if db.crate_graph()[krate].edition < Edition::Edition2021 {
+            if db.crate_graph()[krate].edition < Edition::CURRENT {
```
Comment (lnicola) [inline crates/hir-ty/src/method_resolution.rs:1160]: Shouldn't this stay as `Edition2021`?

Comment (Veykril) [reply crates/hir-ty/src/method_resolution.rs:1160]: Good catch, I misread the comment

**Comment by Urhengulas** [comment]
🙏🏾

**Pull request merged**